funnel/models/post.py

```diff
@@ -0,0 +1,260 @@
+    commentset = db.relationship(
+        Commentset,
+        uselist=False,
+        lazy='joined',
+        cascade='all',
+        single_parent=True,
+        backref=db.backref('post', uselist=False),
+    )
+
+    search_vector = db.deferred(
+        db.Column(
+            TSVectorType(
+                'name',
+                'title',
+                'body_text',
+                weights={'name': 'A', 'title': 'A', 'body_text': 'B'},
+                regconfig='english',
+                hltext=lambda: db.func.concat_ws(' / ', Post.title, Post.body_html,),
+            ),
+            nullable=False,
+        )
+    )
+
+    __table_args__ = (
+        db.CheckConstraint(
+            db.case([(profile_id.isnot(None), 1)], else_=0)
+            + db.case([(project_id.isnot(None), 1)], else_=0)
+            == 1,
+            name='post_owner_check',
+        ),
+    )
+
+    __roles__ = {
+        'reader': {
+            'read': {
+                'body',
+                'created_at',
+                'edited_at',
+                'name',
+                'title',
+                'user',
+                'visibility',
+            }
+        }
+    }
+
+    __datasets__ = {
+        'primary': {
+            'body',
+            'created_at',
+            'edited_at',
+            'name',
+            'title',
+            'user',
+            'visibility',
+        }
+    }
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.voteset = Voteset(settype=SET_TYPE.POST)
+        self.commentset = Commentset(settype=SET_TYPE.POST)
+
+    def __repr__(self):
+        return '<Post "{title}" {uuid_b58}'.format(
+            title=self.title, uuid_b58=self.uuid_b58
+        )
+
+    state.add_conditional_state(
+        'UNPUBLISHED',
+        state.DRAFT,
+        lambda post: post.published_at is not None,
+        lambda post: post.published_at.isnot(None),
+        label=('unpublished', __("Unpublished")),
+    )
+
+    @with_roles(call={'author', 'profile_admin', 'project_editor'})
+    @state.transition(
+        state.DRAFT,
+        state.PUBLISHED,
+        title=__("Publish post"),
+        message=__("Post has been published"),
+    )
+    def publish(self, actor):
+        self.published_by = actor
+        if self.published_at is None:
+            self.published_at = db.func.utcnow()
+
+    @with_roles(call={'author', 'profile_admin', 'project_editor'})
+    @state.transition(
+        state.PUBLISHED,
+        state.DRAFT,
+        title=__("Undo publish"),
+        message=__("Post is now a draft"),
+    )
+    def undo_publish(self):
+        pass
+
+    @with_roles(call={'author', 'profile_admin', 'project_editor'})
+    @state.transition(
+        None,
+        state.DELETED,
+        title=__("Delete post"),
+        message=__("Post has been deleted"),
+    )
+    def delete(self, actor):
+        self.deleted_by = actor
+        self.deleted_at = db.func.utcnow()
```

Comment:
We need hard delete when the `from_` state is a draft (and not _unpublished_). Ideally this should be in the `delete` method itself, so that there is a single API for callers.